docs/explanations/decisions/0007-subpackage-structure.md

```diff
@@ -0,0 +1,82 @@
+- `ophyd_async.tango`: Tango specific signals and devices
+- `ophyd_async.panda`: PandA Device written on top of Epics or Tango modules
+
+This ADR proposes a public sub-package structure. The internal private structure should be flat, but can change according to the number of classes in a public package.
+
+## Decision
+
+### core
+
+There will be a flat public namespace under core, with contents reimported from an underscore prefixed python files, e.g.:
+- `_status.py` for `AsyncStatus`, `WatchableAsyncStatus`, etc.
+- `_protocol.py` for `AsyncReadable`, `AsyncStatus`, etc.
+- `_device.py` for `Device`, `DeviceVector`, etc.
+- `_signal.py` for `Signal`, `SignalBackend`, `observe_signal`, etc.
+- `_readable.py` for `StandardReadable`, `ConfigSignal`, `HintedSignal`, etc.
+- `_detector.py` for `StandardDetector`, `DetectorWriter`, `DetectorControl`, `TriggerInfo`, etc.
+- `_flyer.py` for `StandardFlyer`, `FlyerControl`, etc.
+
+There are some renames that will be required, e.g. `HardwareTriggeredFlyable` -> `StandardFlyer`
+
+### epics
+
+Epics modules consist of 2 sorts of classes:
+- `IO` classes, which are `Device` subclasses containing `Signals` that map as closely as possible to the EPICS template hierarchy. Their name matches the EPICS template with the suffix `IO`.
+- `Control` and `Writer` classes that plug into `StandardDetector` and `StandardFlyer` telling them which logic to use during various bluesky verbs
+
+There should be an additional level of packages that correspond to the epics support module, and contain classes that map to the EPICS database and logic classes e.g.:
+
+- `epics.signal``: containing `epics_signal_rw`, `EpicsTransport`, `CaSignalBackend`, etc.
+- `epics.adcore`: containing `ADDriverIO`, `NDFileHdfIO`, `ADHdfWriter` etc. 
+- `epics.adpilatus`: containing `ADPilatusIO`, `PilatusControl` etc.
+
+The name of the module is the EPICS module lowercased with dashes and other special characters converted to underscores.
+
+They can be imported from modules:
+
+```python
+from ophyd_async.epics import adcore, adpilatus
+
+drv = adpilatus.ADPilatusIO(prefix + "DRV:")
+hdf = adcore.NDFileHDFIO(prefix + "HDF:")
+```
+
+The structure is left to an individual module's size. For instance `motor` will probably be a single `motor.py`, while `ADCore` will likely be an `ADCore/` directory with `_io.py`, `_writer.py`, `_control.py`.
+
+Detector modules should include a reference `StandardDetector` subclass, but without making any site specific decisions about PV naming.
+
+### tango
+
+This has not been created at the time of writing, but it is envisioned that it will follow the EPICS structure where it makes sense.
+
+### panda
+
+This should be a top level namespace containing the `CommonPandaBlocks`, `SeqBlock`, `PandaHdfWriter`, `PandaPcapController`, etc. These should be included in files like `_control.py`, `_writer.py`, `_block.py`, `_table.py` and imported into the top level namespace.
```

Comment:
Would it be useful to briefly justify why `PandA` is special enough to be in ophyd-async directly, in contrast to most of the other devices? (I'm sure there's a good reason, I just don't know what it is, and writing the reason in this or a similar ADR would help me).

Reply:
Good question. `PandA` will eventually be one of a class of devices that are implemented in [FastCS](https://github.com/DiamondLightSource/FastCS) so therefore can use either EPICS or Tango as the transport layer.

Thinking about this, should we make a `fastcs` top level directory then? So we have `from ophyd_async.fastcs import panda` then `panda.HDFPanda`? That would leave the namespace free for other FastCS devices. Although even PandA isn't actually a FastCS device yet, so it's not technically true at the moment...

Reply:
Would like very careful dependency management here. A compatibility layer between two libraries should be kept independent (i.e. if you want to use it you should `pip install ophyd-async[fastcs]`)